error_parity/cvxpy_utils.py

```diff
@@ -375,6 +376,25 @@ def group_positive_prediction_rate(group_idx: int):
             if i < j
         ]
 
+    # If average odds, i.e., equal average tpr and fpr across groups
+    elif fairness_constraint == "average_odds":
+
+        # NOTE: (TPR + FPR) / 2
+        def average_tpr_fpr(group_idx: int):
+            group_tpr = groupwise_roc_points_vars[group_idx][1]
+            group_fpr = groupwise_roc_points_vars[group_idx][0]
+
+            return (group_tpr + group_fpr) / 2
+
+        # Add constraints on the absolute difference between group-wos
+        constraints += [
+            cp.abs(
+                average_tpr_fpr(i) - average_tpr_fpr(j)
```

Comment:
If I'm interpreting this correctly, then this metric definition is based on equality of $(TPR+FPR)/2$ between all groups; are you sure @kschweig?

I thought it would be the same as equalized odds but using average to aggregate differences instead of max.

Reply:
This is indeed a **relaxed version** of equalized odds, where instead of the two constraints $P(\hat{Y} \mid A=1, Y=y) = P(\hat{Y} \mid A=0, Y=y) \quad y \in \\{0, 1\\}$ a single constraint $E_{Y}[P(\hat{Y} \mid A=1, Y)] = E_{Y}[P(\hat{Y} \mid A=0, Y)]$ is defining it. The empirical estimator for this constraint is $(\Delta TPR + \Delta FPR) / 2$. This is especially inspired by AIF360 offering the average odds difference as a metric: https://aif360.readthedocs.io/en/stable/modules/generated/aif360.metrics.ClassificationMetric.html#aif360.metrics.ClassificationMetric.average_odds_difference however I have no link to a paper investigating this metric in depth. Loosely speaking it allows to discriminate one group in TPR if the other is equally discriminated in FPR.

Reply:
thanks for the clarification!

It's a bit of a weird metric because if one group has lower TPR and higher FPR than it can seen as zero constraint violation `average_odds=0.0`; but that would just be the worse of both metrics.

So in fact, a value of `0.0` constraint violation is not strictly equalized odds as there are infinite other solutions to this constraint by balancing higher FPR with lower TPR, or lower FPR with higher TPR.